src/swap.ts

```diff
@@ -0,0 +1,291 @@
+import { purry } from './purry';
+import { isString } from './isString';
+import { isArray } from './isArray';
+import { IterableContainer } from './_types';
+
+type isEqual<A, B> = A extends B ? (B extends A ? true : false) : false;
+
+type isLessThan<A extends number, B extends number> = isEqual<A, B> extends true
+  ? false
+  : 0 extends A
+  ? true
+  : 0 extends B
+  ? false
+  : isLessThan<Subtract<A, 1>, Subtract<B, 1>>;
+
+type BuildTuple<
+  L extends number,
+  T extends ReadonlyArray<unknown> = []
+> = T['length'] extends L ? T : BuildTuple<L, [...T, unknown]>;
+
+type Subtract<A extends number, B extends number> = BuildTuple<A> extends [
+  ...infer U,
+  ...BuildTuple<B>
+]
+  ? U['length']
+  : never;
+
+type IsNonNegativeInteger<T extends number> = number extends T
+  ? false
+  : `${T}` extends `-${string}` | `${string}.${string}`
+  ? false
+  : true;
+
+type Absolute<T extends number> = `${T}` extends `-${infer R}`
```

Comment:
You can mix the two checks, but my bigger concern is that this isn't supported in older typescript versions, I think it was added in 4.7 or even 4.8, and we currently want to support a broader range of typescript versions. 

I think the best approach right now so you can start shipping stuff, and so people can use it is to split the `object` case from the array and string cases. I think that object has a much simpler type, and a much simpler impl, so it's an easy win and i'd be happy to see it ship asap. TL;DR - I'd split this function into `swapIndices` and `swapProperties`.